worker/codalabworker/worker.py

```diff
@@ -188,15 +209,62 @@ def _checkin(self):
                     self._exiting = True
                 self._should_upgrade = True
 
+    def _allocate_cpu_and_gpu_sets(self, request_cpus, request_gpus):
+        """
+        Allocate a cpuset and gpuset to assign to a bundle based on given requested resources.
+        Side effects: updates the free sets, self._cpuset_free and self._gpuset_free
+
+        Arguments:
+            request_cpus: integer
+            request_gpus: integer
+
+        Returns a 3-tuple:
+            success: True if returned cpuset and gpuset are valid, False if cannot allocate resources
+            cpuset: Allocated cpuset. Empty set if success is False
+            gpuset: Allocated gpuset. Empty set if success is False
+        """
+        cpuset, gpuset = set(), set()
+
+        with self._resource_lock:
+            if len(self._cpuset_free) < request_cpus or len(self._gpuset_free) < request_gpus:
+                return False, cpuset, gpuset
+
+            for i in range(request_cpus):
+                cpuset.add(self._cpuset_free.pop())
+            for j in range(request_gpus):
+                gpuset.add(self._gpuset_free.pop())
+            return True, cpuset, gpuset
+
+    def _deallocate_cpu_and_sets(self, cpuset, gpuset):
+        """
+        Release held up cpus and gpus
+
+        Re-add cpuset and gpuset back to their respective free sets
+        """
+        with self._resource_lock:
+            self._cpuset_free |= cpuset
+            self._gpuset_free |= gpuset
+
     def _run(self, bundle, resources):
         if self.shared_file_system:
             bundle_path = bundle['location']
         else:
             bundle_path = self._dependency_manager.get_run_path(bundle['uuid'])
+
+        success, cpuset, gpuset = self._allocate_cpu_and_gpu_sets(
+                resources['request_cpus'], resources['request_gpus'])
+
+        if not success: # revert self._cpuset_free and self._gpuset_free in-place
+            logger.debug('Unsuccessful allocation of cpu and gpu sets for bundle %s', bundle['uuid'])
+            self._deallocate_cpu_and_sets(cpuset, gpuset)
```

Comment:
Is it ever possible to get a case where `success` is `False` but `cpuset` and `gpuset` are non-empty?
It looks like that should not be the case but then why do we need this function?

The threading here LGTM otherwise.

Reply:
Ya on second thoughts that `success` may not be necessary. It's probably a remnant of past iterations by now